### PR TITLE
Energy deposition score and application to depletion

### PR DESCRIPTION
**Pull request by drewejohnson**
This PR introduces two features: an energy deposition tally and a related energy helper for the operator

## Energy deposition tally score
Using the separated fission and heating scoring routines from #1329 and the inclusion of fission heating from #1322, the energy deposition score is implemented simply as
``total_heating - fission_heating + modified_fission_q``. Where ``modified_fission_q`` is the recoverable energy from fission minus energy from prompt neutrons. The idea behind this split is that removing the energy from prompt neutrons [and eventually prompt photons] from the recoverable fission energy, we avoid double counting when including the total heating. This allows the energy from neutrons [and eventually photons] to be distributed away from the fission site.

Some features were added to the `openmc::Tabulated1D` and `openmc::Polynomial` classes to make this process a little easier. It is not possible to construct these with `x` and `y` vectors for `Tabulated1D` and the coefficient vector for `Polynomial`. 

This also assumes that nearly all of the fission heating [MT=318] comes from energy that is contained in the recoverable fission energy [MT=458] and is included in the total heating [MT=301] information. 

Below is a table taken by scoring neutron heating, recoverable fission energy, and the new energy deposition quantity in a 2D PWR assembly. Cross section data was taken using the fixes in #1341. Scores have been divided by 1E6.

|material	| energy-deposition	| fission-q-recoverable	| heating |
|-|-|-|-|
fuel |179.1233756191688	| 94.01349137593269	|87.48114580138251
water |1.588209065664372	| 0.0 |1.588209065664372
cladding | 0.06933964196717787 |0.0 |0.06933964196717787

About 99% of the energy is deposited in the fuel. For non-fissile materials, the energy deposition score recovers the normal neutron heating. For the fuel, if you sum the `fission-q-recoverable` and `heating` columns, you get a value that is greater than the `energy-deposition` column. Interestingly, this difference is not equal to the energy deposited in water and clad, ``94.01 + 87.48 - 179.12 ~= 2.37``, ``1.588 + 0.069 ~= 1.65``

This feature supersedes #1272 and improves it with :arrow_down:

## Energy deposition model for depletion

Using this new score, getting the system energy for depletion is very straightforward. An `EnergyScoreHelper` is added that creates a tally that computes the `energy-deposition` score with no filters, e.g. across the entire problem. This value can be passed directly to the `Operator` to scale reaction rates accordingly. A special case is taken when running with multiple processes in that only the master process returns the tallied data. Since all processes share the same tally data, having each process return the energy produced by "this process" would result in scaling the system energy by the number of MPI processes
https://github.com/openmc-dev/openmc/blob/3df4f94fbf990f7b02b18738ca02ee8614273a03/openmc/deplete/operator.py#L633-L635

## TODO
- [ ] Add this score to the tally regression test [before / after #1338 ?]
- [ ] Remove prompt photon energy if we are in coupled neutron / photon transport

### CODE REVIEW

File: src/nuclide.cpp [hunk 1]
```diff
@@ -263,6 +263,54 @@ Nuclide::Nuclide(hid_t group, const std::vector<double>& temperature, int i_nucl
...
+
+    // Check if fission_q and prompt_neutrons are tabulated 1D
+
+    if (tab_recov) {
+      Tabulated1D* neutrons = dynamic_cast<Tabulated1D*>(prompt_neutrons.get());
+      if (nullptr == neutrons) {
+        throw std::runtime_error{"prompt_neutrons not Tabulated1D in " + 
+          object_name(fer_group)};
+      }
+      auto recov_data = tab_recov->y();
```
Comment (paulromano) [inline src/nuclide.cpp]: Using plain `auto` will make a copy. You probably want
```suggestion
      const auto& recov_data = tab_recov->y();
```

File: src/nuclide.cpp [hunk 2]
```diff
@@ -263,6 +263,54 @@ Nuclide::Nuclide(hid_t group, const std::vector<double>& temperature, int i_nucl
...
+      auto recov_x = tab_recov->x();
```
Comment (paulromano) [inline src/nuclide.cpp]:
```suggestion
      const auto& recov_x = tab_recov->x();
```

File: src/nuclide.cpp [hunk 3]
```diff
@@ -263,6 +263,54 @@ Nuclide::Nuclide(hid_t group, const std::vector<double>& temperature, int i_nucl
...
+        ++i;
+      }
+      modified_fission_q_ = std::make_unique<Tabulated1D>(tab_recov->x(), mod_data);
+    } else {
+        Polynomial* neutrons = dynamic_cast<Polynomial*>(prompt_neutrons.get());
+      if (nullptr == neutrons) {
+        throw std::runtime_error{"prompt_neutrons not Polynomial in " + 
+          object_name(fer_group)};
+      }
+      auto recov_data = poly_recov->coeffs();
```
Comment (paulromano) [inline src/nuclide.cpp]:
```suggestion
      const auto& recov_data = poly_recov->coeffs();
```

File: src/nuclide.cpp [hunk 4]
```diff
@@ -263,6 +263,54 @@ Nuclide::Nuclide(hid_t group, const std::vector<double>& temperature, int i_nucl
...
+      auto neutron_data = neutrons->coeffs();
```
Comment (paulromano) [inline src/nuclide.cpp]:
```suggestion
      const auto& neutron_data = neutrons->coeffs();
```

File: src/endf.cpp [hunk 5]
```diff
@@ -151,6 +151,16 @@ Tabulated1D::Tabulated1D(hid_t dset)
   n_pairs_ = x_.size();
 }
 
+Tabulated1D::Tabulated1D(std::vector<double> x, std::vector<double> y)
+{
+  x_ = x;
+  y_ = y;
```
[the remark this answers is not in the document]
Comment (drewejohnson) [reply src/endf.cpp]: :+1: Incorporated the change. Will push later

**Comment by drewejohnson** [comment]
Superseded by #1350

**Pull request closed without merging**